src/openloco/windows/industry_list.cpp

```diff
@@ -1,16 +1,1330 @@
+        }
+
+        // 0x0045848A
+        static void on_tool_update(window& self, const widget_index widgetIndex, int16_t x, const int16_t y)
+        {
+            registers regs;
+            regs.esi = (int32_t)&self;
+            regs.dx = widgetIndex;
+            regs.ax = x;
+            regs.bx = y;
+            call(0x0045848A, regs);
+        }
+
+        // 0x0045851F
+        static void on_tool_down(window& self, const widget_index widgetIndex, int16_t x, const int16_t y)
+        {
+            registers regs;
+            regs.esi = (int32_t)&self;
+            regs.dx = widgetIndex;
+            regs.ax = x;
+            regs.bx = y;
+            call(0x0045851F, regs);
+        }
+
+        // 0x00458C09
+        static void sub_458C09()
+        {
+            registers regs;
+            call(0x00458C09, regs);
+        }
+
+        // 0x00468FFE hide_gridlines
+        static void hideGridlines()
+        {
+            _gridlines_state--;
+            if (_gridlines_state == 0)
+            {
+                auto window = WindowManager::getMainWindow();
+                if (window != nullptr)
+                {
+                    if ((config::get().flags & config::flags::gridlines_on_landscape) == 0)
+                    {
+                        if ((window->viewports[0]->flags & viewport_flags::gridlines_on_landscape) == 0)
+                        {
+                            window->invalidate();
+                        }
+                        window->viewports[0]->flags ^= viewport_flags::gridlines_on_landscape;
+                    }
+                }
+            }
+        }
+
+        // 0x004585AD
+        static void on_tool_abort(window& self, const widget_index widgetIndex)
+        {
+            sub_458C09();
+            hideGridlines();
+        }
+
+        // 0x0045845F
+        static void on_close(window* self)
+        {
+            if (input::is_tool_active(self->type, self->number))
+                input::cancel_tool();
+        }
+
+        static void init_events()
+        {
+            events.draw = draw;
+            events.draw_scroll = draw_scroll;
+            events.event_08 = event_08;
+            events.on_tool_update = on_tool_update;
+            events.on_tool_down = on_tool_down;
+            events.get_scroll_size = get_scroll_size;
+            events.on_mouse_up = on_mouse_up;
+            events.on_update = on_update;
+            events.scroll_mouse_down = on_scroll_mouse_down;
+            events.scroll_mouse_over = on_scroll_mouse_over;
+            events.prepare_draw = prepare_draw;
+            events.tooltip = tooltip;
+            events.on_tool_abort = on_tool_abort;
+            events.on_close = on_close;
+        }
+    }
+
+    namespace common
+    {
+        struct TabInformation
+        {
+            widget_t* widgets;
+            const widx widgetIndex;
+            window_event_list* events;
+            const uint64_t* enabledWidgets;
```

Comment:
I would copy the enabledWidgets rather than make a pointer. Its just a small amount of more memory and reduces dereferencing.

Reply:
do you mean like this?
```
struct TabInformation
{
    widget_t* widgets;
    const widx widgetIndex;
    window_event_list* events;
    const uint64_t enabledWidgets;
};

static TabInformation tabInformationByTabOffset[] = {
    { industry_list::widgets, widx::tab_industry_list, &industry_list::events, industry_list::enabledWidgets },
    { new_industries::widgets, widx::tab_new_industry, &new_industries::events, new_industries::enabledWidgets },
};
```

Reply:
Yup looks good